File: src/clicketysplit/detection/labeling.py
"""Segment typing and auto-labeling.

Ports the batch label pass of the original lab tool
(``stim_pipeline/segment_recording.py:classify_and_label``): classify
segments by duration, then assign stimulus names to blocked repetitions by
clustering word candidates on inter-token gaps. The Python implementation
and the TS port in the review UI share
``tests/labeling_test_vectors.json`` so they stay in sync.

``LabelAnchor`` lives in :mod:`.base` (re-exported here for callers that
prefer to import everything labeling-related from this module).
"""

from __future__ import annotations

from typing import Literal

from .base import LabelAnchor, LabeledSegment, ProposedSegment

__all__ = [
    "LabelAnchor",
    "auto_label",
    "classify_segments",
]


# ---------------------------------------------------------------------------
# Typing
# ---------------------------------------------------------------------------

# Long segments are crosstalk. The original scales the cutoff off the
# recording's own median word length rather than trusting the fixed ceiling,
# so a slow speaker's ordinary words don't get thrown out
# (segment_recording.py:623).
LONG_SEGMENT_FACTOR = 2.2


def _median(values: list[float]) -> float:
    if not values:
        return 0.0
    ordered = sorted(values)
    mid = len(ordered) // 2
    if len(ordered) % 2:
        return ordered[mid]
    return (ordered[mid - 1] + ordered[mid]) / 2.0


def _percentile(values: list[float], pct: float) -> float:
    """Linear-interpolated percentile matching ``numpy.percentile``'s default."""
    if not values:
        return 0.0
    ordered = sorted(values)
    pos = (len(ordered) - 1) * (pct / 100.0)
    lo = int(pos)
    hi = min(lo + 1, len(ordered) - 1)
    frac = pos - lo
    return ordered[lo] * (1.0 - frac) + ordered[hi] * frac


def _long_segment_threshold(
    durations: list[float], min_word_duration_ms: int, max_word_duration_ms: int
) -> float:
    """Crosstalk cutoff adapted to this recording's typical word length.

    Port of segment_recording.py:614-623: estimate a typical word from the
    interquartile band (so crosstalk can't drag the estimate up), then take
    ``max(max_word_duration_ms, median * LONG_SEGMENT_FACTOR)``.
    """
    if not durations:
        return float(max_word_duration_ms)
    p25 = _percentile(durations, 25)
    p75 = _percentile(durations, 75)
    lo = max(p25, float(min_word_duration_ms))
    word_like = [d for d in durations if lo <= d <= p75]
    median_dur = _median(word_like) if word_like else _median(durations)
    return max(float(max_word_duration_ms), median_dur * LONG_SEGMENT_FACTOR)


def _detect_intro_block(segments: list[ProposedSegment]) -> int:
    """Return the number of leading segments that look like an intro block.

    The legacy pipeline's heuristic: scan the first few inter-segment gaps,
    find the largest, and treat it as an intro/recording-start boundary if
    it is markedly larger than the median later gap. Only used when the
    caller passes ``drop_intro_block=True``.
    """
    n_seg = len(segments)
    if n_seg < 3:
        return 0

    gaps: list[float] = []
    for i in range(1, n_seg):
        gaps.append(segments[i].start - segments[i - 1].end)

    if len(gaps) < 2:
        return 0

    # Limit scan to the first ~8 segments (intro is at most a few utterances).
    scan_range = max(3, min(8, n_seg // 10))
    early_gaps = gaps[:scan_range]
    if not early_gaps:
        return 0
    max_early_idx = max(range(len(early_gaps)), key=lambda i: early_gaps[i])
    max_early_gap = early_gaps[max_early_idx]
    later_gaps = gaps[scan_range:] if len(gaps) > scan_range else gaps
    if not later_gaps:
        return 0
    sorted_later = sorted(later_gaps)
    median_later = sorted_later[len(sorted_later) // 2]

    if max_early_gap > median_later * 1.5 and max_early_gap > 0.6:
        return max_early_idx + 1
    return 0


def classify_segments(
    segments: list[ProposedSegment],
    *,
    min_word_duration_ms: int = 500,
    max_word_duration_ms: int = 1400,
    drop_intro_block: bool = True,
) -> list[LabeledSegment]:
    """Type-classify segments by duration, optionally flagging an intro block.

    ``short_noise`` < ``min_word_duration_ms``; ``crosstalk`` above the
    adaptive long-segment threshold (``max_word_duration_ms`` scaled up when
    the recording's median word runs long — see
    :func:`_long_segment_threshold`); ``word`` in between.
    ``assigned_name`` and ``label_source`` are left blank here —
    :func:`auto_label` fills them in for word-typed segments.
    """
    intro_count = _detect_intro_block(segments) if drop_intro_block else 0

    long_thresh = _long_segment_threshold(
        [s.duration_ms for s in segments[intro_count:]],
        min_word_duration_ms,
        max_word_duration_ms,
    )

    out: list[LabeledSegment] = []
    for i, seg in enumerate(segments):
        if i < intro_count:
            out.append(
                LabeledSegment(
                    start=seg.start,
                    end=seg.end,
                    duration_ms=seg.duration_ms,
                    segment_type="intro",
                    assigned_name="",
                    label_source="",
                    status="intro",
                )
            )
            continue

        if seg.duration_ms < min_word_duration_ms:
            stype: Literal["word", "short_noise", "crosstalk", "intro"] = "short_noise"
        elif seg.duration_ms > long_thresh:
            stype = "crosstalk"
        else:
            stype = "word"

        out.append(
            LabeledSegment(
                start=seg.start,
                end=seg.end,
                duration_ms=seg.duration_ms,
                segment_type=stype,
                assigned_name="",
                label_source="",
                status="pending",
            )
        )
    return out


# ---------------------------------------------------------------------------
# Auto-labeling (forward walk from anchors)
# ---------------------------------------------------------------------------


def _word_indices(segments: list[LabeledSegment]) -> list[int]:
    """Absolute segment indices of word-typed segments, in order."""
    return [i for i, s in enumerate(segments) if s.segment_type == "word"]


def _validate_anchors(
    anchors: list[LabelAnchor], n_words: int, stimulus_list: list[str]
) -> list[LabelAnchor]:
    """Sort anchors by word_index; drop out-of-range or unknown-label anchors.

    A label of ``""`` is allowed (it halts forward labeling per spec). Any
    other label must be a member of ``stimulus_list`` — otherwise we silently
    drop the anchor, matching the spec's "anchors override at their own
    index" without letting a typo poison the walk.
    """
    valid: list[LabelAnchor] = []
    stim_set = set(stimulus_list)
    for a in anchors:
        if a.word_index < 0 or a.word_index >= n_words:
            continue
        if a.label != "" and a.label not in stim_set:
            continue
        valid.append(a)
    # Stable sort by word_index; if duplicates exist later anchors win.
    valid.sort(key=lambda a: a.word_index)
    deduped: dict[int, LabelAnchor] = {}
    for a in valid:
        deduped[a.word_index] = a
    return sorted(deduped.values(), key=lambda a: a.word_index)


def auto_label(
    segments: list[LabeledSegment],
    stimulus_list: list[str],
    *,
    presentation_order: Literal["random", "cycled", "blocked"],
    expected_reps_per_stimulus: int = 2,
    anchors: list[LabelAnchor] | None = None,
) -> list[LabeledSegment]:
    """Apply tentative auto-labels to word-typed segments.

    Non-word segments are returned unchanged. Per presentation order:

    - ``random``: only anchored word indices get a label
      (``label_source="anchor"``); everything else stays blank.
    - ``cycled``: walk forward stride-1 through ``stimulus_list`` from the
      most recent anchor. Between two anchors, hold the earlier anchor's
      stimulus position until the later anchor's index, then jump.
    - ``blocked``: cluster word tokens on inter-token gaps and assign one
      stimulus per cluster, in list order. ``expected_reps_per_stimulus``
      is a fallback hint, not a hard stride, so a speaker producing an
      extra or missing repetition does not shift every later label. An
      anchor pins its cluster's stimulus; later clusters continue from
      there.

    Anchors with ``label == ""`` halt forward labeling — every word-typed
    segment from that index onward stays blank until the next anchor.

    For ``blocked`` and ``cycled`` with no anchor at ``word_index == 0``,
    an implicit ``LabelAnchor(0, stimulus_list[0], "initial")`` is added.
    """
    segments = [
        LabeledSegment(
            start=s.start,
            end=s.end,
            duration_ms=s.duration_ms,
            segment_type=s.segment_type,
            assigned_name=s.assigned_name,
            label_source=s.label_source,
            status=s.status,
            token_index=s.token_index,
            cluster_size=s.cluster_size,
        )
        for s in segments
    ]

    word_idx = _word_indices(segments)
    n_words = len(word_idx)

    # Reset any previous auto-labels on word-typed segments before re-walking.
    for i in word_idx:
        segments[i].assigned_name = ""
        segments[i].label_source = ""

    if n_words == 0:
        return segments

    anchors = list(anchors) if anchors is not None else []
    anchors = _validate_anchors(anchors, n_words, stimulus_list)

    # Project anchors onto their word-typed segments first; they always win
    # at their own index regardless of presentation_order.
    anchor_by_wi: dict[int, LabelAnchor] = {a.word_index: a for a in anchors}
    for a in anchors:
        seg_idx = word_idx[a.word_index]
        segments[seg_idx].assigned_name = a.label
        segments[seg_idx].label_source = "anchor"

    if presentation_order == "random":
        # No forward walk; anchored labels above are the entire output.
        return segments

    if not stimulus_list:
        return segments

    # Add an implicit anchor at index 0 if missing — see 03 doc and brief.
    if 0 not in anchor_by_wi:
        implicit = LabelAnchor(word_index=0, label=stimulus_list[0], source="initial")
        anchors = sorted([implicit, *anchors], key=lambda a: a.word_index)
        anchor_by_wi[0] = implicit
        segments[word_idx[0]].assigned_name = implicit.label
        segments[word_idx[0]].label_source = "anchor"

    if presentation_order == "cycled":
        _walk_cycled(segments, word_idx, anchors, stimulus_list)
    elif presentation_order == "blocked":
        _walk_blocked(
            segments, word_idx, anchors, stimulus_list, expected_reps_per_stimulus
        )
    else:
        raise ValueError(f"Unknown presentation_order: {presentation_order!r}")

    return segments


def _next_anchor_index(anchors: list[LabelAnchor], after_wi: int) -> int | None:
    """Return the word_index of the first anchor strictly after ``after_wi``."""
    for a in anchors:
        if a.word_index > after_wi:
            return a.word_index
    return None


def _walk_cycled(
    segments: list[LabeledSegment],
    word_idx: list[int],
    anchors: list[LabelAnchor],
    stimulus_list: list[str],
) -> None:
    n_words = len(word_idx)
    n_stim = len(stimulus_list)

    for a_pos, a in enumerate(anchors):
        if a.label == "":
            # Empty anchor halts; everything up to the next anchor stays "".
            continue
        try:
            base_stim_idx = stimulus_list.index(a.label)
        except ValueError:
            continue

        next_a_wi = anchors[a_pos + 1].word_index if a_pos + 1 < len(anchors) else None

        for wi in range(a.word_index + 1, n_words):
            if next_a_wi is not None and wi >= next_a_wi:
                break
            if next_a_wi is not None:
                # Between two anchors: hold the earlier anchor's stimulus
                # rather than stride forward. The next anchor will set the
                # label at its own index. Matches the cycled worked example.
                segments[word_idx[wi]].assigned_name = a.label
                segments[word_idx[wi]].label_source = "auto"
            else:
                stride = wi - a.word_index
                segments[word_idx[wi]].assigned_name = stimulus_list[
                    (base_stim_idx + stride) % n_stim
                ]
                segments[word_idx[wi]].label_source = "auto"


def _gap_clusters(
    segments: list[LabeledSegment],
    word_idx: list[int],
    n_stim: int,
    k: int,
) -> list[list[int]]:
    """Split word positions into ``n_stim`` clusters at the largest gaps.

    Speakers pause briefly between repetitions of the same word and longer
    when moving to the next one, so the ``n_stim - 1`` widest gaps are the
    word-switch points. Falls back to fixed ``k``-sized chunks when there
    are too few gaps to split on.

    Returns clusters of *word positions* (indices into ``word_idx``).
    """
    n_words = len(word_idx)
    n_splits = n_stim - 1
    if n_words == 0:
        return []
    if n_splits <= 0:
        return [list(range(n_words))]

    gaps: list[tuple[float, int]] = []
    for wi in range(1, n_words):
        gap = segments[word_idx[wi]].start - segments[word_idx[wi - 1]].end
        gaps.append((gap, wi))

    if len(gaps) < n_splits:
        k = max(1, int(k))
        return [list(range(i, min(i + k, n_words))) for i in range(0, n_words, k)]

    widest = sorted(gaps, key=lambda g: g[0], reverse=True)[:n_splits]
    split_positions = sorted(wi for _, wi in widest)

    clusters: list[list[int]] = []
    prev = 0
    for sp in split_positions:
        clusters.append(list(range(prev, sp)))
        prev = sp
    clusters.append(list(range(prev, n_words)))
    return [c for c in clusters if c]


def _cluster_of(clusters: list[list[int]], word_position: int) -> int | None:
    """Index of the cluster containing ``word_position``, or None."""
    for c_idx, cluster in enumerate(clusters):
        if word_position in cluster:
            return c_idx
    return None


def _walk_blocked(
    segments: list[LabeledSegment],
    word_idx: list[int],
    anchors: list[LabelAnchor],
    stimulus_list: list[str],
    k: int,
) -> None:
    """Label blocked repetitions by clustering on inter-token gaps.

    Consuming exactly ``k`` tokens per stimulus drifts irrecoverably as soon
    as a speaker produces more or fewer repetitions than the script called
    for — every downstream label shifts. Instead we infer cluster boundaries
    from the recording's own timing (see :func:`_gap_clusters`), so ``k`` acts
    as the hint it was always meant to be.

    Anchors still win: an anchor pins its whole cluster to that stimulus, and
    later clusters continue forward through ``stimulus_list`` from there.
    """
    n_words = len(word_idx)
    n_stim = len(stimulus_list)
    if n_words == 0 or n_stim == 0:
        return

    clusters = _gap_clusters(segments, word_idx, n_stim, k)
    if not clusters:
        return

    anchored_positions = {a.word_index for a in anchors if a.label != ""}

    # An anchor re-pins the stimulus for its cluster; clusters after it walk
    # forward from that new position.
    anchor_by_cluster: dict[int, int] = {}
    for a in anchors:
        if a.label == "" or a.label not in stimulus_list:
            continue
        c_idx = _cluster_of(clusters, a.word_index)
        if c_idx is not None:
            anchor_by_cluster[c_idx] = stimulus_list.index(a.label)

    stim_idx = anchor_by_cluster.get(0, 0)
    for c_idx, cluster in enumerate(clusters):
        if c_idx in anchor_by_cluster:
            stim_idx = anchor_by_cluster[c_idx]

        name = stimulus_list[stim_idx] if 0 <= stim_idx < n_stim else ""
        for token_pos, wp in enumerate(cluster):
            seg = segments[word_idx[wp]]
            seg.assigned_name = name
            seg.label_source = "anchor" if wp in anchored_positions else "auto"
            seg.token_index = token_pos + 1
            seg.cluster_size = len(cluster)

        stim_idx += 1

    # Empty-label anchors halt forward labeling: every word from the anchor's
    # index up to the next anchor stays blank (see LabelAnchor in base.py and
    # the auto_label docstring). Applied last so the cluster assignment above
    # cannot overwrite the halted range.
    for a_pos, a in enumerate(anchors):
        if a.label != "":
            continue
        halt_end = (
            anchors[a_pos + 1].word_index if a_pos + 1 < len(anchors) else n_words
        )
        for wp in range(a.word_index, halt_end):
            seg = segments[word_idx[wp]]
            seg.assigned_name = ""
            seg.label_source = "anchor" if wp == a.word_index else ""
            seg.token_index = 0
            seg.cluster_size = 0

File: frontend/src/lib/labeling.ts
// TypeScript port of clicketysplit.detection.labeling.auto_label.
//
// This module is the contract-bound counterpart to the Python labeling
// engine in src/clicketysplit/detection/labeling.py. The shared test corpus
// lives at tests/labeling_test_vectors.json at the repo root; both
// implementations must produce identical output for every vector.
//
// Anchors are keyed by `word_index` — the zero-based position among
// WORD-TYPED segments. Non-word segments (`short_noise`, `crosstalk`,
// `intro`) are never re-typed or relabeled by this routine.
//
// If the algorithm here ever disagrees with the JSON vector file, the
// bug is in this port — match the Python output.
//
// This file is intentionally pure (no DOM, no fetch, no Svelte runes) so
// it can run in vitest, in a Web Worker, or in the main UI without setup.

import type { LabelAnchor, PresentationOrder, Segment } from "./types";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Return the absolute segment indices of all word-typed segments, in
 * document order. The list's length is the count of word-typed segments,
 * and `word_index` anchors index into it.
 */
function wordIndices(segments: readonly Segment[]): number[] {
  const out: number[] = [];
  for (let i = 0; i < segments.length; i++) {
    if (segments[i].segment_type === "word") out.push(i);
  }
  return out;
}

/**
 * Sort, range-check, dedupe, and stimulus-validate anchors.
 *
 * - Anchors with `word_index` outside [0, n_words) are silently dropped.
 * - Anchors with a non-empty label that is not in `stimulus_list` are
 *   silently dropped (match the Python implementation; a typo doesn't
 *   poison the walk).
 * - Duplicate word_index entries: later ones win, matching the Python
 *   `deduped: dict[int, LabelAnchor]` pass.
 */
function validateAnchors(
  anchors: readonly LabelAnchor[],
  nWords: number,
  stimulusList: readonly string[],
): LabelAnchor[] {
  const stim = new Set(stimulusList);
  const dedup = new Map<number, LabelAnchor>();
  for (const a of anchors) {
    if (a.word_index < 0 || a.word_index >= nWords) continue;
    if (a.label !== "" && !stim.has(a.label)) continue;
    dedup.set(a.word_index, a);
  }
  return Array.from(dedup.values()).sort((a, b) => a.word_index - b.word_index);
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

export interface AutoLabelOptions {
  presentationOrder: PresentationOrder;
  expectedRepsPerStimulus?: number;
  anchors?: readonly LabelAnchor[];
}

/**
 * Re-label word-typed segments using a forward walk from anchors.
 *
 * Returns a NEW array of segment objects; the input is not mutated. Only
 * `assigned_name` and `label_source` are touched on each output segment;
 * other fields (start/end/status/...) are copied through.
 *
 * Mirrors Python `auto_label(segments, stimulus_list, *,
 * presentation_order, expected_reps_per_stimulus=2, anchors=None)`.
 */
export function autoLabel(
  segments: readonly Segment[],
  stimulusList: readonly string[],
  options: AutoLabelOptions,
): Segment[] {
  // Shallow-clone every segment so callers' inputs stay untouched.
  const out: Segment[] = segments.map((s) => ({ ...s }));

  const wIdx = wordIndices(out);
  const nWords = wIdx.length;

  // Reset auto-label state on every word-typed segment before walking.
  for (const i of wIdx) {
    out[i].assigned_name = "";
    out[i].label_source = "";
  }

  if (nWords === 0) return out;

  const validatedAnchors = validateAnchors(
    options.anchors ?? [],
    nWords,
    stimulusList,
  );

  // Project anchors onto their segments first; they always win regardless
  // of presentation order.
  const anchorByWi = new Map<number, LabelAnchor>();
  for (const a of validatedAnchors) {
    anchorByWi.set(a.word_index, a);
    out[wIdx[a.word_index]].assigned_name = a.label;
    out[wIdx[a.word_index]].label_source = "anchor";
  }

  if (options.presentationOrder === "random") {
    return out;
  }

  if (stimulusList.length === 0) {
    return out;
  }

  // For cycled/blocked, add an implicit anchor at word_index 0 if none.
  let anchors = validatedAnchors;
  if (!anchorByWi.has(0)) {
    const implicit: LabelAnchor = {
      word_index: 0,
      label: stimulusList[0],
      source: "initial",
    };
    anchors = [implicit, ...anchors].sort(
      (a, b) => a.word_index - b.word_index,
    );
    anchorByWi.set(0, implicit);
    out[wIdx[0]].assigned_name = implicit.label;
    out[wIdx[0]].label_source = "anchor";
  }

  if (options.presentationOrder === "cycled") {
    walkCycled(out, wIdx, anchors, stimulusList);
  } else if (options.presentationOrder === "blocked") {
    walkBlocked(
      out,
      wIdx,
      anchors,
      stimulusList,
      options.expectedRepsPerStimulus ?? 2,
    );
  }

  return out;
}

// ---------------------------------------------------------------------------
// cycled / blocked walks
// ---------------------------------------------------------------------------

function walkCycled(
  segments: Segment[],
  wIdx: readonly number[],
  anchors: readonly LabelAnchor[],
  stimulusList: readonly string[],
): void {
  const nWords = wIdx.length;
  const nStim = stimulusList.length;

  for (let aPos = 0; aPos < anchors.length; aPos++) {
    const a = anchors[aPos];
    if (a.label === "") continue; // empty anchor halts forward labeling

    const baseStimIdx = stimulusList.indexOf(a.label);
    if (baseStimIdx < 0) continue;

    const nextAWi =
      aPos + 1 < anchors.length ? anchors[aPos + 1].word_index : null;

    for (let wi = a.word_index + 1; wi < nWords; wi++) {
      if (nextAWi !== null && wi >= nextAWi) break;
      const segIdx = wIdx[wi];
      if (nextAWi !== null) {
        // Between two anchors: hold the earlier anchor's label.
        segments[segIdx].assigned_name = a.label;
        segments[segIdx].label_source = "auto";
      } else {
        const stride = wi - a.word_index;
        segments[segIdx].assigned_name =
          stimulusList[((baseStimIdx + stride) % nStim + nStim) % nStim];
        segments[segIdx].label_source = "auto";
      }
    }
  }
}

/**
 * Split word positions into `nStim` clusters at the largest gaps.
 *
 * Port of Python `_gap_clusters` (labeling.py), itself a transcription of the
 * original tool's classify_and_label: speakers pause briefly between
 * repetitions of the same word and longer when switching words, so the
 * `nStim - 1` widest gaps are the word-switch points. Falls back to fixed
 * `k`-sized chunks when there are too few gaps to split on.
 *
 * Returns clusters of word positions (indices into `wIdx`). Both sorts are
 * stable in JS and Python, so tie behavior on equal gaps matches too.
 */
function gapClusters(
  segments: readonly Segment[],
  wIdx: readonly number[],
  nStim: number,
  k: number,
): number[][] {
  const nWords = wIdx.length;
  const nSplits = nStim - 1;
  if (nWords === 0) return [];
  const range = (lo: number, hi: number): number[] =>
    Array.from({ length: hi - lo }, (_, i) => lo + i);
  if (nSplits <= 0) return [range(0, nWords)];

  const gaps: Array<[number, number]> = [];
  for (let wi = 1; wi < nWords; wi++) {
    gaps.push([segments[wIdx[wi]].start - segments[wIdx[wi - 1]].end, wi]);
  }

  if (gaps.length < nSplits) {
    const kk = Math.max(1, Math.floor(k));
    const chunks: number[][] = [];
    for (let i = 0; i < nWords; i += kk) {
      chunks.push(range(i, Math.min(i + kk, nWords)));
    }
    return chunks;
  }

  const widest = [...gaps].sort((a, b) => b[0] - a[0]).slice(0, nSplits);
  const splitPositions = widest.map((g) => g[1]).sort((a, b) => a - b);

  const clusters: number[][] = [];
  let prev = 0;
  for (const sp of splitPositions) {
    clusters.push(range(prev, sp));
    prev = sp;
  }
  clusters.push(range(prev, nWords));
  return clusters.filter((c) => c.length > 0);
}

/** Index of the cluster containing `wordPosition`, or null. */
function clusterOf(clusters: number[][], wordPosition: number): number | null {
  for (let cIdx = 0; cIdx < clusters.length; cIdx++) {
    if (clusters[cIdx].includes(wordPosition)) return cIdx;
  }
  return null;
}

/**
 * Label blocked repetitions by clustering on inter-token gaps.
 *
 * Line-for-line port of Python `_walk_blocked` (labeling.py). Consuming
 * exactly `k` tokens per stimulus drifts irrecoverably when a speaker
 * produces more or fewer repetitions than scripted, so cluster boundaries
 * come from the recording's own timing and `k` is only the fallback hint.
 * Anchors pin their whole cluster's stimulus; later clusters continue
 * forward from there. Empty-label anchors blank their range through to the
 * next anchor.
 */
function walkBlocked(
  segments: Segment[],
  wIdx: readonly number[],
  anchors: readonly LabelAnchor[],
  stimulusList: readonly string[],
  k: number,
): void {
  const nWords = wIdx.length;
  const nStim = stimulusList.length;
  if (nWords === 0 || nStim === 0) return;

  const clusters = gapClusters(segments, wIdx, nStim, k);
  if (clusters.length === 0) return;

  const anchoredPositions = new Set(
    anchors.filter((a) => a.label !== "").map((a) => a.word_index),
  );

  // An anchor re-pins the stimulus for its cluster; clusters after it walk
  // forward from that new position.
  const anchorByCluster = new Map<number, number>();
  for (const a of anchors) {
    if (a.label === "" || !stimulusList.includes(a.label)) continue;
    const cIdx = clusterOf(clusters, a.word_index);
    if (cIdx !== null) anchorByCluster.set(cIdx, stimulusList.indexOf(a.label));
  }

  let stimIdx = anchorByCluster.get(0) ?? 0;
  for (let cIdx = 0; cIdx < clusters.length; cIdx++) {
    const pinned = anchorByCluster.get(cIdx);
    if (pinned !== undefined) stimIdx = pinned;

    const name = stimIdx >= 0 && stimIdx < nStim ? stimulusList[stimIdx] : "";
    const cluster = clusters[cIdx];
    for (let tokenPos = 0; tokenPos < cluster.length; tokenPos++) {
      const seg = segments[wIdx[cluster[tokenPos]]];
      seg.assigned_name = name;
      seg.label_source = anchoredPositions.has(cluster[tokenPos])
        ? "anchor"
        : "auto";
      seg.token_index = tokenPos + 1;
      seg.cluster_size = cluster.length;
    }

    stimIdx++;
  }

  // Empty-label anchors halt forward labeling: every word from the anchor's
  // index up to the next anchor stays blank. Applied last so the cluster
  // assignment above cannot overwrite the halted range.
  for (let aPos = 0; aPos < anchors.length; aPos++) {
    const a = anchors[aPos];
    if (a.label !== "") continue;
    const haltEnd =
      aPos + 1 < anchors.length ? anchors[aPos + 1].word_index : nWords;
    for (let wp = a.word_index; wp < haltEnd; wp++) {
      const seg = segments[wIdx[wp]];
      seg.assigned_name = "";
      seg.label_source = wp === a.word_index ? "anchor" : "";
      seg.token_index = 0;
      seg.cluster_size = 0;
    }
  }
}

// ---------------------------------------------------------------------------
// Convenience: project anchors out of segments for round-tripping
// ---------------------------------------------------------------------------

/**
 * Helper for tests / round-trip checks: count word-typed segments before
 * the given absolute segment index. Returns the `word_index` you would
 * assign to a NEW anchor on this segment. Returns -1 if the segment is
 * not word-typed.
 */
export function wordIndexOf(
  segments: readonly Segment[],
  segmentIdx: number,
): number {
  if (segmentIdx < 0 || segmentIdx >= segments.length) return -1;
  if (segments[segmentIdx].segment_type !== "word") return -1;
  let count = 0;
  for (let i = 0; i < segmentIdx; i++) {
    if (segments[i].segment_type === "word") count++;
  }
  return count;
}
